### DISCUSSION TITLE
JSON Parse from Microsoft Graph API

### DISCUSSION THREAD

**Discussion by ghost** [grafana/grafana-infinity-datasource#619] [Q&A]
My goal is display a list of upcoming Client Secret expirations in Grafana (and hopefully build alerts). I've managed to Infinity connected to the Microsoft Graph API and I am getting the JSON Payload, sample below. I set my root to Value, and all the Apps load in per row. I set columns to AppID and then try to extract passwordCredentials.endDateTime but it returns nothing. I've tried type JSON / GRAPHQL and Parser of Default and Backend without any luck. I am not sure what I am missing? The passwordCredentials can also have multiple entries, the example below only has one.

I also saw version 2.0 was released and installed that without luck (was previously 1.4.1).

This is my selector for the column: passwordCredentials.endDateTime

`
{

    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications(id,appId,displayName,passwordCredentials)",

    "value": [        
           {
            "id": "xxxxxxxxxx,
            "appId": "yyyyyyyyyyyy",
            "displayName": "some-application",
            "passwordCredentials": [
                {
                    "customKeyIdentifier": null,
                    "displayName": null,
                    "endDateTime": "2024-08-18T14:43:02.8427168Z",
                    "hint": "NaS",
                    "keyId": "bbbbbbbbaaaaaaaaaaa",
                    "secretText": null,
                    "startDateTime": "2022-08-18T14:43:02.8427168Z"
                }
            ]
        },`

![image](https://github.com/yesoreyeram/grafana-infinity-datasource/assets/130491282/c981e1b5-f1ac-4f03-88a5-fbfef482f16a)

**Comment by ghost** [comment]
One of our developers took a look at it this morning, and they came up with this passwordCredentials.0.endDateTime, which pulls the the first value from the passwordCredentials array. The issue is now how do we iterate through multiple secrets per app without having to statically create multiple columns. Ideally it would be one on each row.

**Reply by yesoreyeram** [reply]
Looks like you needed to switch to backend parser and use JSONata or UQL parser with mv-expand command which iterates over all the values.
